lib/ExecutionEngine/ExecutionEngine.cpp

```diff
@@ -371,3 +376,26 @@ llvm::Error ExecutionEngine::invoke(StringRef name,
 
   return llvm::Error::success();
 }
+
+void MLIRObjectCache::notifyObjectCompiled(const llvm::Module *module,
+                                           llvm::MemoryBufferRef objBuffer) {
+  cachedObjects[module->getModuleIdentifier()] =
+      llvm::MemoryBuffer::getMemBufferCopy(objBuffer.getBuffer(),
+                                           objBuffer.getBufferIdentifier());
+}
+
+void MLIRObjectCache::dumpToObjectFile(llvm::StringRef outputFilename) {
+  // Set up the output file.
+  std::string errorMessage;
+  auto file = openOutputFile(outputFilename, &errorMessage);
+  if (!file) {
+    llvm::errs() << errorMessage << "\n";
+    return;
+  }
+
+  // Dump the object generated for a single module to the output file. 
+  assert(cachedObjects.size() == 1 && "Expected only one object entry.");
```

Comment:
The execution engine API does not preclude from reusing the same object cache for different engines/runs without cleaning it.  Since the cache has a module id-based map, just use the module id to retrieve the correct buffer.

Reply:
The problem is that the cache requires an `llvm::Module` and we call this function from clients that don't have access to it (e.g., JitRunner). Only `mlir::ModuleOp`. If there is no other way to retrieve the `llvm::Module` generated for a given `mlir::ModuleOp`, I could introduce a map `mlir::ModuleOp` -> `llvm::Module` in `SimpleObjectCache`. However, I'm not sure if this would be a problem in an scenario where we generate multiple `llvm::Module`s for a single `mlir::ModuleOp`.